src/token.js

```diff
@@ -1,33 +1,50 @@
 import {isValidElement, cloneElement} from 'react';
 
+
 export const defaultToToken = (candidate, i) => {
   // establish the candidate is valid
   const isObject = !!candidate && (typeof candidate === 'object');
-  const keys = isObject ? Object.keys(candidate) : [];
-  const isValid = (keys.length === 1);
+  const fields = isObject ? Object.keys(candidate) : [];
+  const isValid = (fields.length === 1);
+  const field = isValid && fields[0];
 
-  // determine key and name
-  const key = isValid ? keys[0] : i;
-  const name = isValid ? `__${key}__` : String(candidate);
+  // label is human readable, name is for msgid, key i for React
+  //  names can possibly repeat but it costs us nothing to ensure the label and key are more robust
+  const label = isValid ? `${i}:${field}` : i;
+  const name = isValid ? `__${field}__` : String(candidate);
+  const key = isValid ? `${field}-${i}` : i;
 
   // react elements must have a valid key
-  const pending = isValid ? candidate[key] : candidate;
+  const pending = isValid ? candidate[field] : candidate;
   const value = isValidElement(pending) && !pending.key ?
     cloneElement(pending, {key}) :
     pending;
 
-  return {name, key, value};
+  return {label, name, key, value};
 };
 
+
+export const calculateCollisions = tokens =>
+  tokens
+    .map((token, i, arr) => {
+      const collidesWith = arr.slice(i + 1)
+        .map(other => ((other.name === token.name) && (other.value !== token.value) && other))
+        .filter(Boolean);
+      return !!collidesWith.length &&
+        [token].concat(collidesWith)
+          .map(x => x.label)
+          .join(' vs ');
+    })
+    .filter(Boolean)
+    .filter((msg, i, arr) =>
+      !arr.slice(0, i).find(prev => (prev.indexOf(msg) >= 0))
+    );
+
+
 export const assertTokens = (tokens, message) => {
-  const errors = tokens
-    .filter((token, i, arr) =>
-      arr.slice(i + 1).some(following =>
-        (following.key === token.key) && (following.value !== token.value)
-      ))
-    .map(token => token.key);
-
-  if (errors.length) {
-    throw new Error(`${message}: substitution must be unique or have unique key: ${errors}`);
+  const collisions = calculateCollisions(tokens);
+  if (collisions.length) {
+    throw new Error(
+      `${message}: substitution with the same bane must have the same value: ${collisions}`);
```

Comment:
I don't understand why same name substitutions must have the same value. Why do you care about it?

```js
gettext`bla=${{bla: 'bla'}} foo=${{bla: 'foo'}}`

// would produce  --> `bla=__bla__ foo=__bla__` template
// but then replace into `bla=bla foo=foo` string

Though I may miss some implementation details and this is not actually possible...